### ISSUE TITLE
Setting timeout explicitly

### ISSUE THREAD

**Issue by bukem** [panodata/grafana-client#78]
As of now timeout is set to 5 seconds and there is no way to change it. Please add this option to `from_url `method.

Thanks and regards, Mikhail

**Comment by amotl** [comment]
Dear Mikhail,

thanks a stack for suggesting this. We will bring it in on the next development iteration. Patches are always welcome.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Hi again,

sorry again for my late response.

I've just checked the code, and found that `GrafanaApi()` already obtains an argument `timeout`, which unfortunately defaults to 5 seconds. I agree with Mikhail that this should also be available as an argument to `from_url()`, amongst other improvements in the same area, and - most importantly - this option should have proper documentation, right?

In the meanwhile, and with any current release of `grafana-client`, you can do:
```python
# Create Grafana API wrapper instance, and set timeout.
grafana = GrafanaApi.from_url("https://username:password@daq.example.org/grafana/")
grafana.client.timeout = 60
```

Let me know if that helps you already, so I can handle that with a corresponding priority if it is still a pain, or if that workaround will fit your needs already. Apologies, Mikhail, for missing that detail because I did not have a look at the code yet at the time I was responding to you.

With kind regards,
Andreas.

https://github.com/panodata/grafana-client/blob/63d67567086a2427bd4386a6578907df9dcbe2b3/grafana_client/api.py#L38-L49

**Comment by amotl** [comment]
For the improvements to the implementation, when it will be happening, I would like to leave this note here. The future implementation should closely adhere to the specifications of the underlying HTTP library, `urllib3`, see [^1]. There is also excellent documentation about this topic at [^2].

Following the `urllib3` implementation, the new default timeout value will implicitly use `socket.getdefaulttimeout()`, which is `None` by default, at least on my machine. Effectively, this means "infinite" timeout by default.

[^1]: https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html#urllib3.util.Timeout
[^2]: https://datagy.io/python-requests-timeouts/

**Comment by amotl** [comment]
### Choosing sensible default timeout values

To contradict my previous statement, I think `grafana-client` _should_ keep sensible default values for the timeout options, following the recommendation of the `requests` authors.

> Most requests to external servers should have a timeout attached, in case the server is not responding in a timely manner. By default, requests do not time out unless a timeout value is set explicitly. Without a timeout, your code may hang for minutes or more. [^1]

I think following [HTTPX]' recommendations would be a reasonable approach.

> HTTPX defaults to including reasonable timeouts for all network operations, meaning that if a connection is not properly established then it should always raise an error rather than hanging indefinitely. The default timeout for network inactivity is five seconds. [^2]

Effectively, this would mean the default timeout value does not change at all, will be 5 seconds as before, but its configuration interface will be more friendly.

[^1]: https://requests.readthedocs.io/en/latest/user/advanced/#timeouts
[^2]: https://www.python-httpx.org/quickstart/#timeouts

[HTTPX]: https://github.com/encode/httpx

**Comment by amotl** [comment]
Hi again,

corresponding improvements have been made with GH-99, and released on behalf of version 3.6.0. Does this satisfy your needs?

With kind regards,
Andreas.
